processor/src/lib.rs

```diff
@@ -496,31 +510,96 @@ where
         self.chiplets.get_mem_value(ctx, addr)
     }
 
-    pub fn into_parts(self) -> (System, Decoder, Stack, RangeChecker, Chiplets, A) {
+    pub fn into_parts(self) -> (System, Decoder, Stack, RangeChecker, Chiplets, H) {
         (
             self.system,
             self.decoder,
             self.stack,
             self.range,
             self.chiplets,
-            self.advice_provider,
+            self.host.into_inner(),
         )
     }
 }
 
+// PROCESS STATE
+// ================================================================================================
+
+/// A trait that defines a set of methods which allow access to the state of the process.
+pub trait ProcessState {
+    /// Returns the current clock cycle of a process.
+    fn clk(&self) -> u32;
+
+    /// Returns the current execution context ID.
+    fn ctx(&self) -> u32;
+
+    /// Returns the value located at the specified position on the stack at the current clock cycle.
+    fn get_stack_item(&self, pos: usize) -> Felt;
+
+    /// Returns a word located at the specified word index on the stack.
+    ///
+    /// Specifically, word 0 is defined by the first 4 elements of the stack, word 1 is defined
+    /// by the next 4 elements etc. Since the top of the stack contains 4 word, the highest valid
+    /// word index is 3.
+    ///
+    /// The words are created in reverse order. For example, for word 0 the top element of the
+    /// stack will be at the last position in the word.
+    ///
+    /// Creating a word does not change the state of the stack.
+    fn get_stack_word(&self, word_idx: usize) -> Word;
+
+    /// Returns stack state at the specified clock cycle. This includes the top 16 items of the
+    /// stack + overflow entries.
+    ///
+    /// # Panics
+    /// Panics if invoked for non-last clock cycle on a stack instantiated with
+    /// `keep_overflow_trace` set to false.
+    fn get_stack_state_at(&self, clk: u32) -> Vec<Felt>;
+
+    /// Returns a word located at the specified context/address, or None if the address hasn't
+    /// been accessed previously.
+    fn get_mem_value(&self, ctx: u32, addr: u32) -> Option<Word>;
+}
+
+impl<H: Host> ProcessState for Process<H> {
+    fn clk(&self) -> u32 {
+        self.system.clk()
+    }
+
+    fn ctx(&self) -> u32 {
+        self.system.ctx()
+    }
+
+    fn get_stack_item(&self, pos: usize) -> Felt {
+        self.stack.get(pos)
+    }
+
+    fn get_stack_word(&self, word_idx: usize) -> Word {
+        self.stack.get_word(word_idx)
+    }
+
+    fn get_stack_state_at(&self, clk: u32) -> Vec<Felt> {
+        self.stack.get_state_at(clk)
+    }
+
+    fn get_mem_value(&self, ctx: u32, addr: u32) -> Option<Word> {
+        self.get_memory_value(ctx, addr)
```

Comment:
I would probably replace this line with `self.chiplets.get_mem_value(ctx, addr)` and delete `get_memory_value()` method on line 509 above..